### PR TITLE
Rwa-155: Implements Role Permissions for GET task

### PR DESCRIPTION
**Pull request by alexbaeza**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-155

### Change description ###

Added Check for task permissions with integration with role assignment service.

Created AccessControlService, PermissionEvaluatorService and Role managementService.
Minor changes to Camunda service logic to handle feign exceptions in private methods making it more re-usable.
Added some optimisations and comments to describe them.
Added domain entities.
Rename CcdIdGenerator to CaseId.
Extended Permission evaluator to check for attributes conditionally too. Added enums for Camunda Variables Added enums for role attributes Updating tests
Fix timestamp assertion on functional tests. Refactored camunda common assertion to use the RestApiActions Remove unused imports.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/permission/PermissionEvaluatorService.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import static uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.camunda.CamundaVariableDefinition.JURISDICTION;
+import static uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.camunda.CamundaVariableDefinition.LOCATION;
+import static uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.camunda.CamundaVariableDefinition.REGION;
+import static uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.camunda.CamundaVariableDefinition.SECURITY_CLASSIFICATION;
+
+@Service
+@SuppressWarnings({
+        "java:S3776",
+        "PMD.DataflowAnomalyAnalysis", "PMD.LawOfDemeter", "PMD.AvoidDeeplyNestedIfStmts", "PMD.CyclomaticComplexity"
+    })
```
Comment (alexbaeza) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/permission/PermissionEvaluatorService.java:34]: These suppressions were specifically added due to: 

- The class contains quite a fair bit of logic and business rules.
- The class was written to make the logic easy to follow and debug.

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java [hunk 2]
```diff
@@ -2,87 +2,220 @@
...
-            authorizationHeadersProvider.getLawFirmAAuthorization()
+            authorizationHeadersProvider.getTribunalCaseworkerAAuthorization()
         );
 
         result.then().assertThat()
             .statusCode(HttpStatus.NOT_FOUND.value())
             .and()
             .contentType(APPLICATION_JSON_VALUE)
-            .body("timestamp", is(notNullValue()))
+            .body("timestamp", is(instanceOf(String.class)))
```
Comment (iguannago) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java:41]: that change is better than before. 
Another suggestion to validate dates  could be  something like: 
```
assertThat(Timestamp.valueOf(text)).isNotNull();
 assertThat(LocalDateTime.parse(text)).isNotNull();
```

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/role/entities/Assignment.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@SuppressWarnings("PMD.ExcessiveParameterList")
+public class Assignment {
+
+    private UUID id;
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/role/entities/Assignment.java:20]: Suggestion: I am aware the role-assignment team defines this as UUID. However, in our case, I don't see why we need to keep UUID and carry with it rather than just use a simple String.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/role/entities/request/RoleAssignmentRequest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
+package uk.gov.hmcts.reform.wataskmanagementapi.auth.role.entities.request;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.google.common.base.Objects;
+import uk.gov.hmcts.reform.wataskmanagementapi.auth.role.entities.Assignment;
+
+import java.util.List;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/role/entities/request/RoleAssignmentRequest.java:9]: from what I can tell, I think this is the default Jackson object mapper configuration in spring-boot. 
See here: https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto-customize-the-jackson-objectmapper

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/clients/RoleAssignmentServiceApi.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    )
+    GetRoleAssignmentResponse getRolesForUser(@PathVariable("user-id") String userId,
+                                              @RequestHeader(AUTHORIZATION) String userToken,
+                                              @RequestHeader(SERVICE_AUTHORIZATION) String serviceAuthToken);
+
+    @PostMapping(
+        value = "/am/role-assignments/query",
+        produces = "application/vnd.uk.gov.hmcts.role-assignment-service.post-assignment-query-request+json;charset=UTF-8;version=1.0"
+    )
+    Token token(@RequestBody Map<String, ?> form);
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/clients/RoleAssignmentServiceApi.java:38]: Don't understand this endpoint? does it exist? Where are the authentication headers?

**Review by iguannago — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/role/entities/Assignment.java (hunk 3 above)
Comment (alexbaeza) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/role/entities/Assignment.java:20]: Thanks I'll apply suggestion

**Review by iguannago — DISMISSED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/EndpointSecurityTest.java [hunk 6]
```diff
@@ -131,7 +130,7 @@ public void should_return_401_when_bearer_token_is_provided_but_no_service_token
 
         given()
             .relaxedHTTPSValidation()
-            .header(authorizationHeadersProvider.getLawFirmAAuthorizationOnly())
+            .header(authorizationHeadersProvider.getCaseworkerAAuthorizationOnly())
             .contentType(MediaType.APPLICATION_JSON_VALUE)
             .baseUri(testUrl)
             .pathParam("task-id", taskId)
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/EndpointSecurityTest.java:136]: We already have claim above. Should this be cancel? not part of this PR but just noticed.

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java [hunk 7]
```diff
@@ -2,87 +2,227 @@
...
+        result.then().assertThat()
+            .statusCode(HttpStatus.OK.value())
+            .and().contentType(MediaType.APPLICATION_JSON_VALUE)
+            .and().body("task.id", equalTo(task.get("taskId")));
     }
+
+    @Test
+    public void should_return_a_403_when_user_did_not_have_permission_jurisdiction_region_location_did_not_match() {
+        Map<String, String> task = common.setupTaskAndRetrieveIdsWithCustomVariable(LOCATION, "17595");
+
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java:208]: Should this have a customVariable for jurisdiction, region and location too fulfil the test scenario? Just having a customVariable for LOCATION makes this test behave like "should_return_a_403_when_the_user_did_not_have_sufficient_permission_location_did_not_match()" above.

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java [hunk 8]
```diff
@@ -2,87 +2,227 @@
...
+            .contentType(APPLICATION_JSON_VALUE)
+            .body("timestamp", lessThanOrEqualTo(LocalDateTime.now()
+                .format(DateTimeFormatter.ofPattern(DATE_TIME_FORMAT))))
+            .body("error", equalTo(HttpStatus.FORBIDDEN.getReasonPhrase()))
+            .body("status", equalTo(HttpStatus.FORBIDDEN.value()))
+            .body("message", equalTo(
+                format("User did not have sufficient permissions to access task with id: %s", task.get("taskId"))
+            ));
+    }
+
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java:226]: Should we have tests that capture the 401 and 500 AC's? I.e.

- Then the endpoint will return a 401 HTTP error (not a valid user)

- Then the endpoint will return a 500 HTTP error (role assignment service error)

**Review by BenNzewi — APPROVED** [review]
